File: sdks/rust_sdk/examples/workflow.rs
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Workflow Orchestration Example for SW4RM Protocol.
//!
//! Demonstrates DAG-based workflow execution using the real SDK
//! `WorkflowClient`:
//! - Creating workflow definitions with nodes and dependencies
//! - Starting workflow instances and tracking node states
//! - Advancing nodes through Ready → Running → Completed transitions
//! - Using `resume_workflow` for automatic dependency advancement
//!
//! Execution mode: Local in-memory demo. The `WorkflowClient` validates
//! the DAG, manages node state transitions, and detects completion
//! automatically. No external services are required.
//!
//! Run:
//! ```bash
//! cargo run --example workflow
//! ```

use std::collections::HashMap;
use sw4rm_sdk::clients::workflow::{
    NodeStatus, TriggerType, WorkflowClient, WorkflowDefinition, WorkflowNode,
};

// ============================================================================
// Pipeline Definition
// ============================================================================

/// Create a sample CI/CD pipeline workflow.
///
/// Pipeline structure:
///   build → test → (security_scan, quality_check) → deploy
///
/// security_scan and quality_check run in parallel after test completes.
/// deploy only runs after both complete.
fn create_cicd_pipeline() -> WorkflowDefinition {
    let mut nodes = HashMap::new();

    nodes.insert(
        "build".to_string(),
        WorkflowNode::new("build".to_string(), "build-agent".to_string())
            .unwrap()
            .with_trigger(TriggerType::Manual)
            .with_input("source_repo".to_string(), "source_repo".to_string())
            .with_output("artifact_path".to_string(), "build_artifact".to_string())
            .with_metadata("timeout_ms".to_string(), "300000".to_string()),
    );

    nodes.insert(
        "test".to_string(),
        WorkflowNode::new("test".to_string(), "test-agent".to_string())
            .unwrap()
            .with_dependency("build".to_string())
            .with_trigger(TriggerType::Dependency)
            .with_input("artifact".to_string(), "build_artifact".to_string())
            .with_output("test_results".to_string(), "test_results".to_string())
            .with_output("coverage".to_string(), "coverage".to_string())
            .with_metadata("test_type".to_string(), "unit,integration".to_string()),
    );

    nodes.insert(
        "security_scan".to_string(),
        WorkflowNode::new("security_scan".to_string(), "security-agent".to_string())
            .unwrap()
            .with_dependency("test".to_string())
            .with_trigger(TriggerType::Dependency)
            .with_input("artifact".to_string(), "build_artifact".to_string())
            .with_output(
                "vulnerabilities".to_string(),
                "security_report".to_string(),
            )
            .with_metadata("scan_type".to_string(), "SAST,DAST".to_string()),
    );

    nodes.insert(
        "quality_check".to_string(),
        WorkflowNode::new("quality_check".to_string(), "quality-agent".to_string())
            .unwrap()
            .with_dependency("test".to_string())
            .with_trigger(TriggerType::Dependency)
            .with_input("coverage".to_string(), "coverage".to_string())
            .with_output("quality_score".to_string(), "quality_score".to_string())
            .with_metadata("min_coverage".to_string(), "80".to_string()),
    );

    nodes.insert(
        "deploy".to_string(),
        WorkflowNode::new("deploy".to_string(), "deploy-agent".to_string())
            .unwrap()
            .with_dependency("security_scan".to_string())
            .with_dependency("quality_check".to_string())
            .with_trigger(TriggerType::Dependency)
            .with_input("artifact".to_string(), "build_artifact".to_string())
            .with_input("security".to_string(), "security_report".to_string())
            .with_input("quality".to_string(), "quality_score".to_string())
            .with_output("deployment_url".to_string(), "deployment_url".to_string())
            .with_metadata("environment".to_string(), "staging".to_string()),
    );

    WorkflowDefinition::new("cicd-pipeline-1".to_string(), nodes).unwrap()
}

// ============================================================================
// Node simulation
// ============================================================================

/// Simulate node execution with realistic output.
fn simulate_node_output(node_id: &str) -> String {
    match node_id {
        "build" => serde_json::json!({
            "artifact_path": "/artifacts/build-123.tar.gz",
            "commit_sha": "abc123",
            "build_time_ms": 45000,
        })
        .to_string(),
        "test" => serde_json::json!({
            "test_results": { "passed": 142, "failed": 0, "skipped": 3 },
            "coverage": 87.5,
            "execution_time_ms": 120000,
        })
        .to_string(),
        "security_scan" => serde_json::json!({
            "vulnerabilities": { "critical": 0, "high": 1, "medium": 3, "low": 8 },
            "scan_time_ms": 60000,
        })
        .to_string(),
        "quality_check" => serde_json::json!({
            "quality_score": 92,
            "coverage_met": true,
            "code_smells": 5,
        })
        .to_string(),
        "deploy" => serde_json::json!({
            "deployment_url": "https://staging.example.com",
            "deployment_id": "deploy-456",
            "health_check": "passed",
        })
        .to_string(),
        _ => r#"{"status":"completed"}"#.to_string(),
    }
}

// ============================================================================
// Workflow execution loop
// ============================================================================

/// Run a workflow instance to completion using the SDK client's
/// `resume_workflow` method, which handles dependency advancement.
fn run_workflow(client: &WorkflowClient, instance_id: &str) {
    let node_order = ["build", "test", "security_scan", "quality_check", "deploy"];

    for iteration in 1..=10 {
        let instance = client.get_workflow_status(instance_id).unwrap();
        if instance.is_completed() {
            break;
        }

        // Find ready nodes
        let ready_nodes: Vec<String> = instance
            .node_states
            .iter()
            .filter(|(_, s)| s.status == NodeStatus::Ready)
            .map(|(id, _)| id.clone())
            .collect();

        // Also include Pending root nodes on the first iteration
        let actionable_nodes: Vec<String> = if iteration == 1 {
            instance
                .node_states
                .iter()
                .filter(|(_, s)| s.status == NodeStatus::Pending || s.status == NodeStatus::Ready)
                .map(|(id, _)| id.clone())
                .collect()
        } else {
            ready_nodes.clone()
        };

        if actionable_nodes.is_empty() {
            break;
        }

        println!("\n{}", "=".repeat(60));
        println!("Workflow Iteration {}", iteration);
        println!("{}", "=".repeat(60));

        // Sort by the expected pipeline order for deterministic output
        let mut sorted_nodes = actionable_nodes;
        sorted_nodes.sort_by_key(|n| {
            node_order
                .iter()
                .position(|&x| x == n.as_str())
                .unwrap_or(99)
        });

        println!("Actionable nodes: {}", sorted_nodes.join(", "));

        for node_id in &sorted_nodes {
            // Mark as RUNNING
            client
                .update_node_state(instance_id, node_id, NodeStatus::Running, None, None)
                .unwrap();
            println!("\n[Node] Executing: {}", node_id);

            // Simulate work
            let output = simulate_node_output(node_id);

            // Mark as COMPLETED and advance dependents
            let updated = client
                .resume_workflow(instance_id, node_id, Some(output.clone()), None)
                .unwrap();

            let truncated = if output.len() > 80 {
                format!("{}...", &output[..80])
            } else {
                output
            };
            println!("[Node] Completed: {}", node_id);
            println!("  Output: {}", truncated);

            // Print current state after each node
            println!("\n  Current states:");
            let mut states: Vec<_> = updated.node_states.iter().collect();
            states.sort_by_key(|(id, _)| {
                node_order
                    .iter()
                    .position(|&x| x == id.as_str())
                    .unwrap_or(99)
            });
            for (id, state) in &states {
                println!("    {}: {:?}", id, state.status);
            }
        }
    }
}

// ============================================================================
// Main
// ============================================================================

fn main() {
    println!("SW4RM Workflow Orchestration Example");
    println!("{}", "=".repeat(60));
    println!("\nThis example uses the SDK WorkflowClient to demonstrate");
    println!("DAG-based workflow execution for multi-agent orchestration.\n");

    let client = WorkflowClient::new();

    // Create the CI/CD pipeline workflow
    let definition = create_cicd_pipeline();

    println!("Workflow Structure:");
    println!("  build -> test -> (security_scan, quality_check) -> deploy");
    println!("  * security_scan and quality_check run in parallel");
    println!("  * deploy waits for both to complete\n");

    let workflow_id = client.create_workflow(definition).unwrap();
    println!("Created workflow: {}", workflow_id);

    // Start the workflow
    let instance = client.start_workflow(&workflow_id).unwrap();
    println!("Started instance: {}", instance.instance_id);

    // Run to completion
    run_workflow(&client, &instance.instance_id);

    // Final state
    let final_state = client
        .get_workflow_status(&instance.instance_id)
        .unwrap();

    println!("\n{}", "=".repeat(60));
    println!("WORKFLOW COMPLETE");
    println!("{}", "=".repeat(60));
    println!("\nAll nodes completed: {}", final_state.is_completed());
    println!("Has failures: {}", final_state.has_failures());

    if let (Some(start), Some(end)) = (final_state.started_at, final_state.completed_at) {
        let duration = (end - start).num_milliseconds();
        println!("Duration: {}ms", duration);
    }

    // --------------------------------------------------------------------------
    // Failure Path: Node failure cascades to workflow failure
    // --------------------------------------------------------------------------
    println!("\n{}", "=".repeat(60));
    println!("FAILURE PATH: Node failure cascade");
    println!("{}", "=".repeat(60));

    let fail_def = create_cicd_pipeline();
    let fail_wf_id = client.create_workflow(fail_def).unwrap();
    let fail_instance = client.start_workflow(&fail_wf_id).unwrap();
    println!(
        "\nStarted failure-demo instance: {}",
        fail_instance.instance_id
    );

    // Advance build → Completed
    client
        .update_node_state(&fail_instance.instance_id, "build", NodeStatus::Running, None, None)
        .unwrap();
    client
        .resume_workflow(&fail_instance.instance_id, "build", None, None)
        .unwrap();
    println!("  build: Completed");

    // Mark test as Failed with an error
    client
        .update_node_state(&fail_instance.instance_id, "test", NodeStatus::Running, None, None)
        .unwrap();
    client
        .update_node_state(
            &fail_instance.instance_id,
            "test",
            NodeStatus::Failed,
            None,
            Some("Integration tests timed out".to_string()),
        )
        .unwrap();
    println!("  test: Failed (Integration tests timed out)");

    let fail_final = client
        .get_workflow_status(&fail_instance.instance_id)
        .unwrap();
    println!("\n  has_failures: {}", fail_final.has_failures());
    println!("  Downstream nodes remain Pending:");
    let mut states: Vec<_> = fail_final.node_states.iter().collect();
    states.sort_by_key(|(id, _)| {
        ["build", "test", "security_scan", "quality_check", "deploy"]
            .iter()
            .position(|&x| x == id.as_str())
            .unwrap_or(99)
    });
    for (id, state) in &states {
        println!("    {}: {:?}", id, state.status);
    }

    // Demonstrate cancel_workflow on a fresh instance
    println!("\n--- cancel_workflow ---");
    let cancel_def = create_cicd_pipeline();
    let cancel_wf_id = client.create_workflow(cancel_def).unwrap();
    let cancel_inst = client.start_workflow(&cancel_wf_id).unwrap();
    client.cancel_workflow(&cancel_inst.instance_id).unwrap();
    let cancelled = client
        .get_workflow_status(&cancel_inst.instance_id)
        .unwrap();
    println!(
        "  Cancelled instance: is_completed={}, has_failures={}",
        cancelled.is_completed(),
        cancelled.has_failures()
    );

    println!("\nKey takeaways:");
    println!("1. WorkflowDefinition::new() validates the DAG (cycle detection, dep resolution)");
    println!("2. start_workflow() initializes all nodes as Pending");
    println!("3. resume_workflow() marks a node Complete and auto-promotes dependents to Ready");
    println!("4. The client detects workflow completion when all nodes are terminal");
    println!("5. Parallel execution is possible for independent branches");
    println!("6. A single Failed node marks the workflow as failed; cancel_workflow() fails all non-terminal nodes");
}

File: sdks/ex_sdk/examples/workflow.exs
# SW4RM Elixir SDK — Workflow Orchestration Example
#
# Demonstrates the local DAG-based workflow engine:
# - Defining a multi-node DAG (lint -> test -> build -> deploy)
# - Submitting and tracking workflow progress
# - Advancing through ready nodes
# - Handling node failure and workflow cancellation
# - Status reporting
#
# Run: mix run examples/workflow.exs

alias Sw4rm.Workflow
alias Sw4rm.Workflow.{Node, Edge}

IO.puts("=== SW4RM Workflow Orchestration Example ===\n")

# ──────────────────────────────────────────────
# 1. Start the Workflow Engine
# ──────────────────────────────────────────────
IO.puts("--- 1. Start Workflow Engine ---")

{:ok, engine} = Workflow.start_link()
IO.puts("  Workflow engine started")

# ──────────────────────────────────────────────
# 2. Define a CI/CD Pipeline DAG
# ──────────────────────────────────────────────
IO.puts("\n--- 2. Define CI/CD Pipeline ---")
IO.puts("  DAG: lint ──┐")
IO.puts("               ├──> build ──> deploy")
IO.puts("       test ──┘")

pipeline = %{
  name: "ci-cd-pipeline",
  nodes: [
    %{
      node_id: "lint",
      agent_id: "lint-agent",
      task_type: "lint",
      params: %{strict: true}
    },
    %{
      node_id: "test",
      agent_id: "test-agent",
      task_type: "test",
      params: %{suite: "unit"}
    },
    %{
      node_id: "build",
      agent_id: "build-agent",
      task_type: "build",
      params: %{target: "release"}
    },
    %{
      node_id: "deploy",
      agent_id: "deploy-agent",
      task_type: "deploy",
      params: %{env: "staging"}
    }
  ],
  edges: [
    %{from_node: "lint", to_node: "build"},
    %{from_node: "test", to_node: "build"},
    %{from_node: "build", to_node: "deploy"}
  ]
}

{:ok, wf_id} = Workflow.submit(engine, pipeline)
IO.puts("  Submitted workflow: #{wf_id}")

# ──────────────────────────────────────────────
# 3. Execute the Workflow
# ──────────────────────────────────────────────
IO.puts("\n--- 3. Execute Workflow ---")

# Check initial ready nodes (lint and test have no dependencies)
{:ok, ready} = Workflow.ready_nodes(engine, wf_id)
ready_ids = Enum.map(ready, & &1.node_id) |> Enum.sort()
IO.puts("  Ready nodes: #{inspect(ready_ids)}")

# Complete lint
:ok = Workflow.update_node(engine, wf_id, "lint", :completed, "No lint errors")
IO.puts("  lint: completed (no lint errors)")

# Complete test
:ok = Workflow.update_node(engine, wf_id, "test", :completed, "42 tests passed")
IO.puts("  test: completed (42 tests passed)")

# Now build should be ready
{:ok, ready} = Workflow.ready_nodes(engine, wf_id)
ready_ids = Enum.map(ready, & &1.node_id)
IO.puts("  Ready nodes: #{inspect(ready_ids)}")

# Complete build
:ok = Workflow.update_node(engine, wf_id, "build", :completed, "Artifact v1.2.3 built")
IO.puts("  build: completed (artifact v1.2.3)")

# Deploy should now be ready
{:ok, ready} = Workflow.ready_nodes(engine, wf_id)
ready_ids = Enum.map(ready, & &1.node_id)
IO.puts("  Ready nodes: #{inspect(ready_ids)}")

# Complete deploy
:ok = Workflow.update_node(engine, wf_id, "deploy", :completed, "Deployed to staging")
IO.puts("  deploy: completed (staging)")

# Check final status
{:ok, status} = Workflow.status(engine, wf_id)
IO.puts("\n  Final status: #{status.status}")
IO.puts("    Total nodes: #{status.total_nodes}")
IO.puts("    Completed:   #{status.completed}")
IO.puts("    Pending:     #{status.pending}")
IO.puts("    Failed:      #{status.failed}")

# ──────────────────────────────────────────────
# 4. Failure and Cancellation
# ──────────────────────────────────────────────
IO.puts("\n--- 4. Failure and Cancellation ---")

failure_pipeline = %{
  name: "failing-pipeline",
  nodes: [
    %{node_id: "check", agent_id: "check-agent", task_type: "check", params: %{}},
    %{node_id: "release", agent_id: "release-agent", task_type: "release", params: %{}}
  ],
  edges: [
    %{from_node: "check", to_node: "release"}
  ]
}

{:ok, fail_id} = Workflow.submit(engine, failure_pipeline)
IO.puts("  Submitted failing workflow: #{fail_id}")

# Fail the check node
:ok = Workflow.update_node(engine, fail_id, "check", :failed, "Security scan failed")
IO.puts("  check: failed (security scan)")

{:ok, status} = Workflow.status(engine, fail_id)
IO.puts("  Workflow status after failure: #{status.status}")

# No nodes should be ready since the dependency failed
{:ok, ready} = Workflow.ready_nodes(engine, fail_id)
IO.puts("  Ready nodes after failure: #{inspect(Enum.map(ready, & &1.node_id))}")

# Demonstrate explicit cancellation
cancel_pipeline = %{
  name: "cancel-demo",
  nodes: [
    %{node_id: "step-a", agent_id: "a", task_type: "work", params: %{}},
    %{node_id: "step-b", agent_id: "b", task_type: "work", params: %{}}
  ],
  edges: [
    %{from_node: "step-a", to_node: "step-b"}
  ]
}

{:ok, cancel_id} = Workflow.submit(engine, cancel_pipeline)
:ok = Workflow.cancel(engine, cancel_id)
{:ok, status} = Workflow.status(engine, cancel_id)
IO.puts("\n  Cancellation demo:")
IO.puts("    Submitted and immediately cancelled")
IO.puts("    Status: #{status.status}")

# ──────────────────────────────────────────────
# 5. Listing Workflows
# ──────────────────────────────────────────────
IO.puts("\n--- 5. Listing Workflows ---")

all = Workflow.list(engine)
IO.puts("  Total workflows: #{length(all)}")

for wf <- all do
  IO.puts("    #{wf.name}: #{wf.status} (#{wf.completed}/#{wf.total_nodes} completed)")
end

# ──────────────────────────────────────────────
IO.puts("\n=== Workflow Orchestration Example Complete ===")
IO.puts("\nKey concepts covered:")
IO.puts("  1. DAG-based workflow definition with nodes and edges")
IO.puts("  2. Dependency-aware execution (ready_nodes)")
IO.puts("  3. Node status progression: pending -> completed/failed")
IO.puts("  4. Workflow-level status derived from node statuses")
IO.puts("  5. Explicit cancellation support")
